lib/validateKey.js

```diff
@@ -0,0 +1,70 @@
+const ASYMMETRIC_KEY_DETAILS_SUPPORTED = require('./asymmetricKeyDetailsSupported');
+const RSA_PSS_KEY_DETAILS_SUPPORTED = require('./rsaPssKeyDetailsSupported');
+
+const allowedAlgorithmsForKeys = {
+  'ec': ['ES256', 'ES384', 'ES512'],
+  'rsa': ['RS256', 'PS256', 'RS384', 'PS384', 'RS512', 'PS512'],
+  'rsa-pss': ['PS256', 'PS384', 'PS512'],
+  'ed25519': ['EdDSA'],
+  'ed448': ['EdDSA']
+}
+
+module.exports = function(algorithm, key) {
+  if (!key) return;
+
+  const keyType = key.asymmetricKeyType;
+  if (!keyType) return;
+
+  const allowedAlgorithms = allowedAlgorithmsForKeys[keyType];
+
+  if (!allowedAlgorithms) {
+    throw new Error(`Unknown key type "${keyType}".`);
+  }
+
+  if (!allowedAlgorithms.includes(algorithm)) {
+    throw new Error(`"alg" parameter for "${keyType}" key type must be one of: ${allowedAlgorithms.join(', ')}.`)
+  }
+
+  if (!ASYMMETRIC_KEY_DETAILS_SUPPORTED) return;
+
+  switch (keyType) {
+  case 'ec':
+    const curve = key.asymmetricKeyDetails.namedCurve;
+
+    [
+      { algorithm: 'ES256', curve: 'prime256v1' },
+      { algorithm: 'ES384', curve: 'secp384r1' },
+      { algorithm: 'ES512', curve: 'secp521r1' },
+    ].forEach(function(expected) {
+      if (algorithm === expected.algorithm && curve !== expected.curve) {
+        throw new Error(`"alg" parameter "${algorithm}" requires curve "${expected.curve}".`);
+      }
+    });
+    break;
+
+  case 'rsa':
+    if (key.asymmetricKeyDetails.modulusLength < 2048) {
+      throw new Error(`${algorithm} requires key modulusLength to be 2048 bits or larger.`)
+    }
+    break;
+
+  case 'rsa-pss':
+    if (key.asymmetricKeyDetails.modulusLength < 2048) {
+      throw new Error(`${algorithm} requires key modulusLength to be 2048 bits or larger.`);
+    }
+
```

Comment:
Here are we duplicating the changes made in David's PR? 

https://github.com/auth0/node-jsonwebtoken/pull/852/files#diff-d1ba59a4612dc5aa06926e8794400534698b44a57958a6a6522f642c90ee9223R118-R130

and this would not be covered by the `allowInsecureKeySizes` option.